### PR TITLE
fix(autoscaling): every deployment resets capacity

### LINKED ISSUE #5215: Preserve current desired counts on deploy of auto scaled resources
When configuring an ec2 AutoScalingGroup or ECS service autoscaling (and likely other examples), the CDK requires a desired count value or will set it to a fixed default (1 for ECS, [more complex behaviour](https://github.com/aws/aws-cdk/pull/5208/commits/71321ed84a4dc91cf8aee5357fb8b482d9a37a08) for EC2). In general, this is annoying for autoscaling, as it means that a `cdk deploy` will trample over any actions taken by the scaling policy - possibly causing an outage or overspend, depending on the value set.

I propose that an option is added to these constructs to set the desired count to the current value for the resource if it exists. e.g. if there are 20 instances registered in my autoscaling group at the start of the `cdk deploy`, it should preserve this value.

Implementing this feature would also provide a solution for [this bug report](https://github.com/aws/aws-cdk/issues/3919).

### Use Case

We perform a `cdk deploy` for each of our 10s of ECS services in pipelines, triggered by a code push of our CDK app package. This means that we're doing many frequent deployments, and it's either dangerous or costly to have our resources scale far away from where the scaling policy placed them.

### Proposed Solution

We're working around this issue by querying the current value using the AWS APIs, and explicitly setting the desired count to this value in an attempt to leave it unchanged. My proposal (unless there is a better way) is for the CDK to do this work at deploy time.

Rough Python ECS example:
```python
ecs_client = boto3.client('ecs', region_name=self.region)
existing_services = ecs_client.describe_services(cluster=cluster_name,
                                                 services=[service_name])['services']
if len(existing_services) == 0:
    desired_task_count = 3
else:
    assert len(existing_services) == 1
    desired_task_count = existing_services[0]['desiredCount']

service = ecs.Ec2Service(self, 'EcsService', service_name=service_name, task_definition=task_definition, cluster=ecs_cluster, desired_count=desired_task_count)
scaling = service.auto_scale_task_count(min_capacity=1, max_capacity=100)

scaling.scale_to_track_custom_metric('ServiceAutoScaling',
                                     metric=cloudwatch.Metric(metric_name='Utilization', namespace='ModelServing',
                                                              dimensions={'service': service_name},
                                                              period=core.Duration.minutes(1)),
                                     target_value=0.5)
``` 

We have similar logic for EC2 autoscaling, but it's less generalizable due to not knowing the physical name of the ASG at execution time, but happy to discuss if it's helpful.

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by rix0rrr**
If `DesiredCapacity` is specified in the CloudFormation template, on every
deployment the capacity of the AutoScalingGroup is reset to that number,
even if the group had been scaled out at that point. The solution is to
leave DesiredCapacity empty, in which case it will remain untouched
during a deployment.

Previously, CDK would use some logic to always calculate a
DesiredCapacity for you, even if you left the `desiredCapacity` property
unset, leading to the undesirable behavior--which frankly represents
an availability risk.

Now, if you don't specify `desiredCapacity`, we won't set
`DesiredCapacity` either, avoiding the availability risk that we
introduced beforehand. In fact, if you *do* set `desiredCapacity`, we
will warn you that you probably shouldn't using a construct warning.

Fixes #5215, closes #5208.

BREAKING CHANGE: AutoScalingGroups without `desiredCapacity` are now
initially scaled to their minimum capacity (instead of their maximum
capaciety).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -44,7 +44,10 @@ export interface CommonAutoScalingGroupProps {
   /**
    * Initial amount of instances in the fleet
    *
-   * @default 1
+   * If this is set to a number, every deployment will reset the amount of
+   * instances to this number. It is recommended to leave this value blank.
+   *
+   * @default minCapacity, and leave unchanged during deployment
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Is this accurate, it doesn't look like you're defaulting desiredCapacity

Comment (rix0rrr) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: This describes the end-to-end behavior of what will happen if you leave the value empty.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-as-group-desiredcapacity 

>  If you do not specify a desired capacity, the default is the minimum size of the group.

Comment (eladb) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Add this link in the docstring

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 2]
```diff
@@ -475,30 +478,40 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
...
     });
     withResolved(desiredCapacity, maxCapacity, (desired, max) => {
+      if (desired === undefined) { return; }
       if (max < desired) {
         throw new Error(`Should have desiredCapacity (${desired}) <= maxCapacity (${max})`);
       }
     });
 
+    if (desiredCapacity !== undefined) {
+      this.node.addWarning(`desiredCapacity has been configured. Be aware this will reset the size of your AutoScalingGroup on every deployment.`);
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Provider a link to a github issue in the help message so people can dive deeper/comment.

**Pull request merged**